// angular-admin-lte/esm5/lib/box-info/box-info.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ContentChild } from '@angular/core';
import { BoxInfoContentDirective, BoxInfoFooterDirective, BoxInfoHeaderDirective } from './box-info.directive';
/*
 *
 */
var BoxInfoComponent = /** @class */ (function () {
    function BoxInfoComponent() {
        this.contentStyleClass = 'info-box-number';
        this.footerStyleClass = 'progress-description';
        this.headerStyleClass = 'info-box-text';
        this.iconColor = '#fff';
        this.iconStyleClass = 'ion ion-bag';
        this.styleClass = 'info-box';
    }
    /**
     * @method ngOnInit
     */
    /**
     * \@method ngOnInit
     * @return {?}
     */
    BoxInfoComponent.prototype.ngOnInit = /**
     * \@method ngOnInit
     * @return {?}
     */
    function () {
        if (!this.backgroundColor) {
            this.progressBarBg = this.iconBackgroundColor;
        }
    };
    BoxInfoComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mk-box-info',
                    template: "<div [ngClass]=\"styleClass\" [mkColor]=\"backgroundColor\" mkColorProperty=\"background-color\">\n  <span class=\"info-box-icon\" [mkColor]=\"iconBackgroundColor\" mkColorProperty=\"background-color\">\n    <i [ngClass]=\"iconStyleClass\" [mkFontColor]=\"iconColor\"></i>\n  </span>\n  <div class=\"info-box-content\">\n    <span *ngIf=\"header || boxInfoHeaderDirective\" [ngClass]=\"headerStyleClass\" [mkFontColor]=\"headerColor\">\n      {{header}}\n      <ng-content select=\"mk-box-header\"></ng-content>\n    </span>\n    <span [ngClass]=\"contentStyleClass\" [mkFontColor]=\"contentColor\">\n      <ng-container *ngIf=\"boxInfoHeaderDirective || boxInfoContentDirective || boxInfoFooterDirective; else noDirective\">\n        <ng-content select=\"mk-box-content\"></ng-content>\n      </ng-container>\n      <ng-template #noDirective>\n        <ng-content></ng-content>\n      </ng-template>\n    </span>\n    <div *ngIf=\"progressWidth\" class=\"progress\">\n      <div class=\"progress-bar\" [mkColor]=\"progressBarBg\" mkColorProperty=\"background-color\" [style.width.%]=\"progressWidth\"></div>\n    </div>\n    <span *ngIf=\"footer || boxInfoFooterDirective\" [ngClass]=\"footerStyleClass\" [mkFontColor]=\"footerColor\">\n      {{footer}}\n      <ng-content select=\"mk-box-footer\"></ng-content>\n    </span>\n  </div>\n</div>\n",
                    styles: [".info-box.bg-color>.info-box-content{color:#fff}"]
                }] }
    ];
    BoxInfoComponent.propDecorators = {
        backgroundColor: [{ type: Input }],
        contentStyleClass: [{ type: Input }],
        contentColor: [{ type: Input }],
        footer: [{ type: Input }],
        footerColor: [{ type: Input }],
        footerStyleClass: [{ type: Input }],
        header: [{ type: Input }],
        headerColor: [{ type: Input }],
        headerStyleClass: [{ type: Input }],
        iconBackgroundColor: [{ type: Input }],
        iconColor: [{ type: Input }],
        iconStyleClass: [{ type: Input }],
        progressWidth: [{ type: Input }],
        styleClass: [{ type: Input }],
        boxInfoHeaderDirective: [{ type: ContentChild, args: [BoxInfoHeaderDirective,] }],
        boxInfoFooterDirective: [{ type: ContentChild, args: [BoxInfoFooterDirective,] }],
        boxInfoContentDirective: [{ type: ContentChild, args: [BoxInfoContentDirective,] }]
    };
    return BoxInfoComponent;
}());
export { BoxInfoComponent };
if (false) {
    /** @type {?} */
    BoxInfoComponent.prototype.progressBarBg;
    /** @type {?} */
    BoxInfoComponent.prototype.backgroundColor;
    /** @type {?} */
    BoxInfoComponent.prototype.contentStyleClass;
    /** @type {?} */
    BoxInfoComponent.prototype.contentColor;
    /** @type {?} */
    BoxInfoComponent.prototype.footer;
    /** @type {?} */
    BoxInfoComponent.prototype.footerColor;
    /** @type {?} */
    BoxInfoComponent.prototype.footerStyleClass;
    /** @type {?} */
    BoxInfoComponent.prototype.header;
    /** @type {?} */
    BoxInfoComponent.prototype.headerColor;
    /** @type {?} */
    BoxInfoComponent.prototype.headerStyleClass;
    /** @type {?} */
    BoxInfoComponent.prototype.iconBackgroundColor;
    /** @type {?} */
    BoxInfoComponent.prototype.iconColor;
    /** @type {?} */
    BoxInfoComponent.prototype.iconStyleClass;
    /** @type {?} */
    BoxInfoComponent.prototype.progressWidth;
    /** @type {?} */
    BoxInfoComponent.prototype.styleClass;
    /** @type {?} */
    BoxInfoComponent.prototype.boxInfoHeaderDirective;
    /** @type {?} */
    BoxInfoComponent.prototype.boxInfoFooterDirective;
    /** @type {?} */
    BoxInfoComponent.prototype.boxInfoContentDirective;
}
